src/main/java/com/rackspace/ceres/app/services/QueryService.java

```diff
@@ -29,56 +29,78 @@
 import java.time.Duration;
 import java.time.Instant;
 import java.util.LinkedHashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.data.cassandra.core.cql.ReactiveCqlTemplate;
+import org.springframework.data.redis.core.ReactiveRedisTemplate;
 import org.springframework.stereotype.Service;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 
 @Service
+@Slf4j
 public class QueryService {
 
   private final ReactiveCqlTemplate cqlTemplate;
   private final MetadataService metadataService;
   private final DataTablesStatements dataTablesStatements;
   private final TimeSlotPartitioner timeSlotPartitioner;
   private final AppProperties appProperties;
+  private final ReactiveRedisTemplate<String, List<String>> reactiveRedisTemplate;
 
   @Autowired
   public QueryService(ReactiveCqlTemplate cqlTemplate,
-                      MetadataService metadataService,
-                      DataTablesStatements dataTablesStatements,
-                      TimeSlotPartitioner timeSlotPartitioner,
-                      AppProperties appProperties) {
+      MetadataService metadataService,
+      DataTablesStatements dataTablesStatements,
+      TimeSlotPartitioner timeSlotPartitioner,
+      AppProperties appProperties,
+      ReactiveRedisTemplate reactiveRedisTemplate) {
     this.cqlTemplate = cqlTemplate;
     this.metadataService = metadataService;
     this.dataTablesStatements = dataTablesStatements;
     this.timeSlotPartitioner = timeSlotPartitioner;
     this.appProperties = appProperties;
+    this.reactiveRedisTemplate = reactiveRedisTemplate;
   }
 
-  public Flux<QueryResult> queryRaw(String tenant, String metricName,
+  public Flux<QueryResult> queryRaw(String tenant, String metricName, String metricGroup,
       Map<String, String> queryTags,
       Instant start, Instant end) {
-    return metadataService.locateSeriesSetHashes(tenant, metricName, queryTags)
+    if(StringUtils.isNotEmpty(metricName))  {
+      return getQueryResultFlux(tenant, queryTags, start, end, metricName).checkpoint();
+    } else {
+      return getMetricsFlux(metricGroup)
+          .flatMap(
+              metric -> {
+                return getQueryResultFlux(tenant, queryTags, start, end, metric);
+              }
+          )
+          .checkpoint();
+    }
+  }
+
+  private Flux<QueryResult> getQueryResultFlux(String tenant, Map<String, String> queryTags,
+      Instant start, Instant end, String metricNameWithGroup) {
```

Comment:
metricNameWithGroup you shouldn't have a single parameter with both name and group